### PR TITLE
api: add zone, vm name params in listVmSnapshot response

### LINKED ISSUE #4603: uuids instead of display names are shown in detail view of VM snapshot
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Primate UI
~~~

##### CLOUDSTACK VERSION

~~~
4.15.0
~~~

##### SUMMARY

On details view of an VM snapshot uuids instead of display names for "VM name" and "Zone" are shown

##### STEPS TO REPRODUCE

- create VM snapshot
- open detail view for created snapshot

##### EXPECTED RESULTS

real displayname for "VM name" and "Zone" is shown

##### ACTUAL RESULTS

image

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Adds new params in listVmSnapshot API - `zonename`, `virtualmachinename`

Fixes #4603 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [x] Trivial

### How Has This Been Tested?

UI
After change zone name and VM name are shown
Screenshot from 2021-01-20 17-52-47

CMK:
Before - 
```
> list vmsnapshot vmsnapshotid=db83c1a7-7870-442d-8c27-d9d407a05f11
{
  "count": 1,
  "vmSnapshot": [
    {
      "account": "admin",
      "created": "2021-01-20T17:34:35+0530",
      "current": true,
      "displayname": "vm-snap",
      "domain": "ROOT",
      "domainid": "52611b65-5b16-11eb-b67a-645d8651f45a",
      "hypervisor": "Simulator",
      "id": "db83c1a7-7870-442d-8c27-d9d407a05f11",
      "name": "i-2-3-VM_VS_20210120120435",
      "state": "Ready",
      "tags": [],
      "type": "Disk",
      "virtualmachineid": "3a07d2d4-18ad-4ef4-ac97-7937ba629069",
      "zoneid": "5918c672-47aa-4425-9ff9-d71d79d2b304",
    }
  ]
}
```

After-
```
> list vmsnapshot vmsnapshotid=db83c1a7-7870-442d-8c27-d9d407a05f11
{
  "count": 1,
  "vmSnapshot": [
    {
      "account": "admin",
      "created": "2021-01-20T17:34:35+0530",
      "current": true,
      "displayname": "vm-snap",
      "domain": "ROOT",
      "domainid": "52611b65-5b16-11eb-b67a-645d8651f45a",
      "hypervisor": "Simulator",
      "id": "db83c1a7-7870-442d-8c27-d9d407a05f11",
      "name": "i-2-3-VM_VS_20210120120435",
      "state": "Ready",
      "tags": [],
      "type": "Disk",
      "virtualmachineid": "3a07d2d4-18ad-4ef4-ac97-7937ba629069",
      "virtualmachinename": "test-vm",
      "zoneid": "5918c672-47aa-4425-9ff9-d71d79d2b304",
      "zonename": "Sandbox-simulator"
    }
  ]
}
```

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by harikrishna-patnala — APPROVED** [review]
LGTM

**Comment by shwstppr** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/api/ApiResponseHelper.java [hunk 1]
```diff
@@ -621,11 +621,13 @@ public VMSnapshotResponse createVMSnapshotResponse(VMSnapshot vmSnapshot) {
         vmSnapshotResponse.setDisplayName(vmSnapshot.getDisplayName());
         UserVm vm = ApiDBUtils.findUserVmById(vmSnapshot.getVmId());
         if (vm != null) {
-            vmSnapshotResponse.setVirtualMachineid(vm.getUuid());
+            vmSnapshotResponse.setVirtualMachineId(vm.getUuid());
+            vmSnapshotResponse.setVirtualMachineName(vm.getHostName());
```
Comment (harikrishna-patnala) [inline server/src/main/java/com/cloud/api/ApiResponseHelper.java:625]: I think display_name should be first choice, if it is null then we can go with host_name.

Comment (shwstppr) [reply server/src/main/java/com/cloud/api/ApiResponseHelper.java:625]: okay @harikrishna-patnala

**Comment by shwstppr** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/api/ApiResponseHelper.java (hunk 1 above)
Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/api/ApiResponseHelper.java:627]: Thanks @shwstppr, looks good to me now

**Comment by sureshanaparti** [comment]
@shwstppr I do understand that it is good to display name(s) of resources/entities in the UI. But, I think, to change the respective API responses to include the name(s) of the required resources/entities (to show in the UI) may not be the right thing. The UUIDs can be further used with the list API(s) to get the required details (name, zone, etc) to fill the UI as desired.

I know UI has to attempt multiple API calls, but API responses should primarly return UUIDs. For names and other details, make use of list API calls.

**Comment by rhtyd** [comment]
@shwstppr is this for 4.15 or master, pl change base branch or milestone

**Comment by shwstppr** [comment]
@rhtyd sorry for the error. Changed the base branch

@blueorangutan package

**Comment by shwstppr** [comment]
> @shwstppr I do understand that it is good to display name(s) of resources/entities in the UI. But, I think, to change the respective API responses to include the name(s) of the required resources/entities (to show in the UI) may not be the right thing. The UUIDs can be further used with the list API(s) to get the required details (name, zone, etc) to fill the UI as desired.
> 
> If we continue this exercise to show name(s) in the UI, finally we end up adding name(s) to most of the responses.
> 
> I know UI has to attempt multiple API calls, but API responses should primarly return UUIDs. For names and other details, make use of list API calls.

@sureshanaparti I may agree with you on that partly but changes aren't causing any additional db query, plus we have similar parameters in other api responses so maybe this can be allowed ;)
 Calling list api just for names in UI will be an overkill

**Comment by harikrishna-patnala** [comment]
> > @shwstppr I do understand that it is good to display name(s) of resources/entities in the UI. But, I think, to change the respective API responses to include the name(s) of the required resources/entities (to show in the UI) may not be the right thing. The UUIDs can be further used with the list API(s) to get the required details (name, zone, etc) to fill the UI as desired.
> > If we continue this exercise to show name(s) in the UI, finally we end up adding name(s) to most of the responses.
> > I know UI has to attempt multiple API calls, but API responses should primarly return UUIDs. For names and other details, make use of list API calls.
> 
> @sureshanaparti I may agree with you on that partly but changes aren't causing any additional db query, plus we have similar parameters in other api responses so maybe this can be allowed ;)
> Calling list api just for names in UI will be an overkill

Agree with you @shwstppr

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
> > @shwstppr I do understand that it is good to display name(s) of resources/entities in the UI. But, I think, to change the respective API responses to include the name(s) of the required resources/entities (to show in the UI) may not be the right thing. The UUIDs can be further used with the list API(s) to get the required details (name, zone, etc) to fill the UI as desired.
> > If we continue this exercise to show name(s) in the UI, finally we end up adding name(s) to most of the responses.
> > I know UI has to attempt multiple API calls, but API responses should primarly return UUIDs. For names and other details, make use of list API calls.
> 
> @sureshanaparti I may agree with you on that partly but changes aren't causing any additional db query, plus we have similar parameters in other api responses so maybe this can be allowed ;)
> Calling list api just for names in UI will be an overkill


@shwstppr So, do you mean it is the right approach to consider the API response object parameters for UI display? Few recent PRs below, with such changes where the response parameters are mainly updated for UI display only.

https://github.com/apache/cloudstack/pull/4483 - updated listNetworks API response for VPC Name
https://github.com/apache/cloudstack/pull/4275 - updated listVmSnapshot API response  for Hypervisor
https://github.com/apache/cloudstack/pull/4073 - updated listPublicIpAddress API response  for Network Name

**Comment by sureshanaparti** [comment]
@shwstppr May be some generic solution for the API response objects, to tag both UUID and Name (wherever UUID is returned) might work for all the APIs, and the names can be used appropriately in the UI when required.

**Comment by harikrishna-patnala** [comment]
> @shwstppr May be some generic approach for the API response objects, to tag both UUID and Name (wherever UUID is returned) might work for all the APIs, and the names can be used appropriately in the UI when required.

@sureshanaparti I get your proposal, but that I think it needs significant amount of changes as Name is not part of common interfaces (which I think is required for a generic approach) and it is defined as different names for each resource.

May be an another improvement ticket would help this PR to get continued !

**Comment by rhtyd** [comment]
@blueorangutan ui

**Comment by rhtyd** [comment]
@blueorangutan ui

**Comment by rhtyd** [comment]
@blueorangutan ui

**Comment by rhtyd** [comment]
@blueorangutan ui

**Comment by rhtyd** [comment]
@blueorangutan ui

File: server/src/main/java/com/cloud/api/ApiResponseHelper.java [hunk 2]
```diff
@@ -621,11 +622,14 @@ public VMSnapshotResponse createVMSnapshotResponse(VMSnapshot vmSnapshot) {
         vmSnapshotResponse.setDisplayName(vmSnapshot.getDisplayName());
         UserVm vm = ApiDBUtils.findUserVmById(vmSnapshot.getVmId());
         if (vm != null) {
-            vmSnapshotResponse.setVirtualMachineid(vm.getUuid());
+            vmSnapshotResponse.setVirtualMachineId(vm.getUuid());
+            vmSnapshotResponse.setVirtualMachineName(Strings.isNullOrEmpty(vm.getDisplayName()) ? vm.getHostName() : vm.getDisplayName());
+            vmSnapshotResponse.setVirtualMachineName(vm.getHostName());
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/api/ApiResponseHelper.java:627]: Why is setVirtualMachineName() added twice @shwstppr ?

Comment (shwstppr) [reply server/src/main/java/com/cloud/api/ApiResponseHelper.java]: @rhtyd that was by mistake in my previous commit :facepalm: . Fixed it

**Comment by shwstppr** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/api/ApiResponseHelper.java [hunk 3]
```diff
@@ -621,11 +622,13 @@ public VMSnapshotResponse createVMSnapshotResponse(VMSnapshot vmSnapshot) {
         vmSnapshotResponse.setDisplayName(vmSnapshot.getDisplayName());
         UserVm vm = ApiDBUtils.findUserVmById(vmSnapshot.getVmId());
         if (vm != null) {
-            vmSnapshotResponse.setVirtualMachineid(vm.getUuid());
+            vmSnapshotResponse.setVirtualMachineId(vm.getUuid());
+            vmSnapshotResponse.setVirtualMachineName(Strings.isNullOrEmpty(vm.getDisplayName()) ? vm.getHostName() : vm.getDisplayName());
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/api/ApiResponseHelper.java:626]: LGTM - we need one check if for normal user/account it leaks the internal hostname (is hostname the i-x-y-VM?)

**Review by rhtyd — APPROVED** [review]

File: server/src/main/java/com/cloud/api/ApiResponseHelper.java (hunk 3 above)
Comment (rhtyd) [reply server/src/main/java/com/cloud/api/ApiResponseHelper.java:626]: checked hostname would return name from the Db table, lgtm

**Pull request merged**